### PR TITLE
miranda 2.066 (new formula)

### PR DESCRIPTION
**Pull request by vladimyr**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

I'm aware that [Miranda](http://miranda.org.uk) isn't that popular anymore but it had a strong influence on the Haskell back in the day and after years of being closed source, it went completely open-source recently. I thought it would be cool to add it here so people can install it more easily and give it a try.

/cc @MiroDojkic

### CODE REVIEW

**Comment by Bo98** [comment]
@BrewTestBot test this please

**Comment by vladimyr** [comment]
@Bo98 I figured out why this fails. Miranda compiles stdlib (`$(LIB)/miralib/stdenv.m`) as part of the installation. Produced binary file (`$(LIB)/miralib/stdenv.x`) contains the creation date of its source (`$(LIB)/miralib/stdenv.m`).  Once you run Miranda it will check that those dates match. It works locally but once you put them in the bottle and change all timestamps (https://github.com/Homebrew/brew/blob/10ba0d5/Library/Homebrew/dev-cmd/bottle.rb#L282-L291) they do not match anymore and the only way to fix it is to remove `stdenv.x` and compile it again so it contains updated creation date of `stdenv.m`. Now I'm not sure how to resolve it 🤔 

Why does `brew bottle` alter timestamps anyway? Also is there a way to add some kind of post-install step which would instruct Miranda to recompile stdlib and other stuff that is needed?

**Review by chenrui333 — APPROVED** [review]

File: Formula/miranda.rb
Comment (SMillerDev) [inline Formula/miranda.rb]: Why do these need to be compiled here?

File: Formula/miranda.rb
Comment (SMillerDev) [inline Formula/miranda.rb]: Can you provide a patch upstream to fix this?

File: Formula/miranda.rb
Comment (SMillerDev) [inline Formula/miranda.rb]: Please provide a patch upstream to fix this.

File: Formula/miranda.rb
Comment (SMillerDev) [inline Formula/miranda.rb]: Why is this needed?

**Review by SMillerDev — COMMENTED** [review]
Please patch the source upstream instead of drastically changing the install in the formula.

File: Formula/miranda.rb
Comment (vladimyr) [reply Formula/miranda.rb]: Strictly speaking, examples do not need to be precompiled but miralib aka `stdenv.m` needs to be compiled AOT because it is implicitly imported in any script you are trying to execute. I've excluded compilation from an original makefile by removing [`exfiles` rule](https://github.com/rljacobson/Miranda/blob/49d6dcc/Makefile#L74-L75) due to the issue described in my previous comment: https://github.com/Homebrew/homebrew-core/pull/51003#issuecomment-593660181
Which is why I re-introduced it as part of the post-install script.

File: Formula/miranda.rb
Comment (vladimyr) [reply Formula/miranda.rb]: To reassure no compiled files have been put inside the bottle. It should be sufficient to simply call [`make cleanup`](https://github.com/rljacobson/Miranda/blob/49d6dcc/Makefile#L44-L49), which is already done, but I'm using this an additional safety measure. Again, if you bottle those while altering creation timestamps on accompanying sources you'll get runtime error on the first run of Miranda environment.

**Comment by SMillerDev** [comment]
@homebrew/core any opinions on this compilation on the user machine?

**Comment by Bo98** [comment]
I suppose my question would be: how tasking is it? How long does it take?

**Comment by vladimyr** [comment]
> Please patch the source upstream instead of drastically changing the install in the formula.

I'm not sure that is even possible... 🤔 

I've used the following repo: https://github.com/rljacobson/Miranda/tree/49d6dcc in response to your comments because it contains verbatim import of source code dump that is published on Miranda download page: http://miranda.org.uk/downloads Note that there isn't an actively developed upstream out there and I'm using tarball as a source for this formula. (There are some signs of a continuation of the development under new leadership: https://github.com/ncihnegn/miranda but this is currently a one-man show and I'm not really in favor of using it as upstream/base of the submitted formula atm.)

Miranda wasn't open source until 2020-31-01 and code that has been released was last changed decades ago excluding necessary tweaks to get it to compile using present time compilers. According to bundled readme I could theoretically send a patch to mira-bugs (at) miranda.org.uk but I don't think it will end up accepted.
There are two issues with bundled makefile:
1. Assuming the existence of `BIN`, `LIB` and `MAN` paths which is fairly easy to fix and will probably end up being accepted 🤔 
2. Tweaking *nix permissions and ownership of produced artifacts which does not go well with brew's way of doing things. I'm not really optimistic that Mr. Turner is willing to alter his makefile just to conform to brew's conventions given in mind his ages-old software become vaporware a long time ago. 🤷‍♂

This is a sort of chicken and egg problem. I'm aware of the ugliness of the current solution but I'm proposing it in order to make Miranda (in its current state) more accessible to the general public. Hopefully, that will result in a continuation of development after which upstream should get switched and formula should be cleared out of present hacks.

**Comment by vladimyr** [comment]
> @Homebrew/core any opinions on this compilation on the user machine?

Compilation of Miranda's stdenv?

**Comment by vladimyr** [comment]
Here are benchmark results made on my machine:

image

It ends up around 50ms together with removing stuff from previous compilation so real-world numbers would be even lower.

File: Formula/miranda.rb
Comment (MikeMcQuaid) [inline Formula/miranda.rb]: Why? Can an issue be submitted upstream for this?

File: Formula/miranda.rb
Comment (MikeMcQuaid) [inline Formula/miranda.rb]: Please inline this given it's not dynamic.

File: Formula/miranda.rb
Comment (MikeMcQuaid) [inline Formula/miranda.rb]: Why?

File: Formula/miranda.rb
Comment (MikeMcQuaid) [inline Formula/miranda.rb]: Why?

File: Formula/miranda.rb
Comment (MikeMcQuaid) [inline Formula/miranda.rb]: Why is cleanup needed before install?

File: Formula/miranda.rb
Comment (MikeMcQuaid) [reply Formula/miranda.rb]: Why can't compiled file be put inside the bottle?

Comment (vladimyr) [reply Formula/miranda.rb]: Because of https://github.com/Homebrew/homebrew-core/pull/51003#issuecomment-593660181

**Comment by MikeMcQuaid** [comment]
> Why does `brew bottle` alter timestamps anyway?

To provide a degree of reproducibility (https://reproducible-builds.org).

---

I think we need to find a way to not recompile things in `post_install`. Homebrew is a binary package manager so these sort of hacks unfortunately aren't acceptable in Homebrew/core.

Thanks for you work so far!

File: Formula/miranda.rb
Comment (vladimyr) [reply Formula/miranda.rb]: As I explained earlier upstream in the conventional sense does not exist and this is built upon source code dump that hasn't been actively maintained for a long time.

Comment (vladimyr) [reply Formula/miranda.rb]: As I explained earlier upstream in the conventional sense does not exist and this is built upon source code dump that hasn't been actively maintained for a long time.

File: Formula/miranda.rb
Comment (vladimyr) [reply Formula/miranda.rb]: Because it will end up bottled and ultimately cause failures on the first run due to https://github.com/Homebrew/homebrew-core/pull/51003#issuecomment-593660181

File: Formula/miranda.rb
Comment (MikeMcQuaid) [reply Formula/miranda.rb]: If there is no upstream this software is not maintained and not a good fit for Homebrew/core, sorry (https://docs.brew.sh/Acceptable-Formulae#niche-or-self-submitted-stuff)

File: Formula/miranda.rb
Comment (vladimyr) [reply Formula/miranda.rb]: Because it alters them in a way subsequent `brew remove miranda` attempt fails with insufficient rights.

File: Formula/miranda.rb
Comment (vladimyr) [reply Formula/miranda.rb]: Because tarball, unfortunately, contains some precompiled assets and this clears them out.

File: Formula/miranda.rb
Comment (vladimyr) [reply Formula/miranda.rb]: Just to make it super clear:

>this is built upon source code dump that hasn't been actively maintained for a long time.

It is my assumption. I can't and do not know for sure due to it being closed source until recently.

Maybe the author still occasionally works on it and accepts patches but I'm not willing to start contributing by mailing patches around without any transparency.

Again, I just wanted to make it easier for people to play with it but if this is a bad place to host it I'll move it to personal tap. I'm keeping it open and waiting for your final decision.

Comment (MikeMcQuaid) [reply Formula/miranda.rb]: Ok, will pass on this for now given the above, thanks anyway!

**Comment by vladimyr** [comment]
FWIW I've managed to remove postinstall step from the formula: https://github.com/vladimyr/homebrew-core/blob/b610b25/Formula/miranda.rb

**Comment by FranklinChen** [comment]
Has someone contacted the author of Miranda, David Turner, to resolve these build issues?

**Comment by vladimyr** [comment]
I haven't because these aren't really issues, they are rather consequences of different approaches. Asking to conform to brew's way of doing things didn't sound promising to me in a given context but feel free to try.

**Pull request closed without merging**
